storages/backends/gcloud.py

```diff
@@ -213,9 +242,18 @@ def listdir(self, name):
         if name and not name.endswith('/'):
             name += '/'
 
+        # Iterating manually to use a retry handled version of function
         iterator = self.bucket.list_blobs(prefix=self._encode_name(name), delimiter='/')
-        blobs = list(iterator)
+        actual_iterator = iter(iterator)
+        ensure_next = self.retry_handler(next)
+
+        blobs = []
```

Comment:
I've wrapped the "next" method, because I think that HTTP iterator is lazy and will make API requests only when it's evaluated. The "next" method itself is read-only, so I couldn't replace it with a mock, same as make a mock use any side effects when "next" is called - in this case it returns a list iterator, no matter what's been put there.
That's the situation - list of objects doesn't raise (at least while iterator does not make real requests), because next simply returns the next element of the list, and "next" itself cannot be replaced with a mock, that can raise.

Reply:
Would it be cleaner to simple make your own helper function for listing all the blobs, and just put a single retry wrapper around that? Probably wrapping each next isn't ideal, because then it's hard to control how much timeout you'd want to give it. With a retry handler around next, your timeout is per object, not per method call to this library. But if you make a helper, and just wrap that, you'd have a much more controlled timeout, and hopefully some cleaner code too.